[dynamo] Fix np.issubdtype

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116460
* __->__ #116459

Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

lezcano committed:
[dynamo] Fix np.issubdtype

Fixes the issue described at

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -868,8 +877,21 @@ def call_function(
+                # constant fold
+                return variables.ConstantVariable.create(
+                    self.as_python_constant()(
+                        *[x.as_python_constant() for x in args],
```
```suggestion
                        *(x.as_python_constant() for x in args),
```

lezcano replied:
I copied this from the equivalent TorchVariable code. We should add linter rules if we want these things to be consistent in the codebase

Skylion007 commented on torch/_dynamo/variables/misc.py:
```diff
@@ -833,10 +834,18 @@ class NumpyVariable(VariableTracker):
+    @classmethod
+    def can_constant_fold_through(cls, fn):
+        mod = fn.__module__.split(".")
+        assert len(mod) >= 2 and mod[:2] == ["torch", "_numpy"]
```
```suggestion
        assert len(mod) >= 2
        assert mod[:2] == ["torch", "_numpy"]
```
will make it easier to know which assert failed in the future.

ezyang commented:
Yanbo, could you review the dynamo pieces?

peterbell10 commented on torch/_numpy/_dtypes.py:
```diff
@@ -399,24 +408,22 @@ def issubclass_(arg, klass):
+        return t
+
+    arg1 = str_to_abstract(arg1)
+    arg2 = str_to_abstract(arg2)
```
Is this required? It doesn't appear in `np.issubdtype` and calling `np.issubdtype("integer", np.generic)` raises an error:
```python
    415 if not issubclass_(arg1, generic):
--> 416     arg1 = dtype(arg1).type
    417 if not issubclass_(arg2, generic):
    418     arg2 = dtype(arg2).type

TypeError: data type 'integer' not understood
```

lezcano replied:
The issue is that, to avoid having `np.dtype`s in Dyanmo's FX graph, we "serialize" `np.dtype`s using their strings. This is alright for the most part, as constructions like `np.random.randn(3).astype("float")` work, but here we need to work around this very issue in our implementation of this function.

peterbell10 replied:
`issubdtype` is constant folded though so should never appear directly in an fx graph. Doesn't that mean we can just call it with the original object? Or is the error coming from an `issubdtype` call that's inside another `torch._numpy` call?

lezcano replied:
We could do that, but we would need to reimplement (or better, clean-up) the implementation of NumPy dtypes in dynamo. NumPy dtypes are implemented at the moment via the `as_proxy` call on a NumPy variable:
https://github.com/pytorch/pytorch/blob/3407541b0c58315b6252811eeecc600acf0798b7/torch/_dynamo/variables/misc.py#L834-L848

We could implement them as a `ConstantVariable`, but since we do not want to ever serialize them, I think the current approach is just alright.

peterbell10 replied:
Okay, in that case I think just update this change with a comment on why we differ from numpy and also it would be good to update the code you linked to say that we're using strings to avoid having `np.dtype` objects in the fx graph.

lezcano replied:
done

lezcano committed:
Update on "[dynamo] Fix np.issubdtype"


Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc mruberry rgommers voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

peterbell10 reviewed (approved):
LGTM, modulo Skylion's nits

ezyang commented:
@pytorchbot merge

lezcano committed:
Update on "[dynamo] Fix np.issubdtype"


Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc mruberry rgommers voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

zou3519 commented:
I think this broke CI, see https://github.com/pytorch/pytorch/pull/116459. Seems to be a landrace. Example logs: https://ossci-raw-job-status.s3.amazonaws.com/log/20151601880 , should be sufficient to just remove some expected failures listed in https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/dynamo_test_failures.py#L287

lezcano committed:
Update on "[dynamo] Fix np.issubdtype"


Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc mruberry rgommers voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Fix np.issubdtype"


Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc mruberry rgommers voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "[dynamo] Fix np.issubdtype"


Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc mruberry rgommers voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

davidmerwin commented:
>Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom): ...

@lezcano @pytorchmergebot 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

lezcano committed:
Update on "[dynamo] Fix np.issubdtype"


Fixes the issue described at https://github.com/pytorch/pytorch/issues/93697#issuecomment-1828346590

This doesn't fix the full issue yet, now we hit
```python
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 744, in step
  getattr(self, inst.opname)(inst)
  File
  "/home/lezcano/git/pytorch/pytorch/torch/_dynamo/symbolic_convert.py",
  line 1366, in BUILD_MAP
      assert (
      AssertionError
```

cc mruberry rgommers voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
